bisync: logger, march, and several big bug fixes

nielash opened the pull request:
#### What is the purpose of this change?

A suite of several related fixes tackling bisync's biggest pain points.

Highlights:
- bisync: merge copies and deletes, support `--track-renames` and `--backup-dir` -- fixes #5690 and #5685
- operations: add logger to log list of sync results -- fixes #7282
- sync: report list of synced paths to file -- see #7282
- bisync: generate final listing from sync results, not relisting - fixes #5676
- bisync: rollback listing on sync error
- bisync: generate listings concurrently with march -- fixes #7332
- sync: `--fix-case` flag to rename case insensitive dest -- fixes #4854
- bisync: handle unicode and case normalization consistently - fixes #7270
- bisync: optimize `--resync` performance -- partially addresses #5681
- bisync: add support for Google Docs - fixes #5696

#### Was the change discussed in an issue or in the forum before?

Yes. See [this post](https://forum.rclone.org/t/bisync-bugs-and-feature-requests/37636) and the individual issues linked above.

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

nielash commented:
(I think the windows error is unrelated, possibly a side-effect of [recent Go security fix](https://github.com/golang/go/commit/9e933c189ca3a84f12995b3c799364a06abc4376).)

nielash commented:
@ncw Thought I'd bring https://github.com/rclone/rclone/pull/7410/commits/f65dc5f7191fbebbaf889c45331c4e181ade980e to your attention related to your [fix from yesterday](https://github.com/rclone/rclone/commit/acf1e2df84a350b7a86d7672d749dfb1ba090a44) regarding the `filepath.VolumeName` change in go1.21.4. The same issue appeared to be breaking `NewFs` on Windows in a way that wasn't previously showing up on tests. (I added one here that catches it.)

ncw commented on cmd/lsf/lsf_test.go:
```diff
@@ -141,7 +141,7 @@ file3
 	var expectedOutput string
 	for _, item := range items {
-		expectedOutput += item.ModTime(context.Background()).Format("2006-01-02 15:04:05") + "\n"
+		expectedOutput += item.ModTime(context.Background()).Format("2006-01-02 15:04:05.000000000") + "\n"
```
This is a potentially backwards incompatible change.

I think `lsf` attempts to show the correct number of digits for the Precision in the Fs.

nielash replied:
I don't think it does, unless there's an option I'm missing:

```
rclone lsf a --format stp --separator ' '
5 2023-11-17 01:33:40 hi copy.txt
5 2023-11-17 01:30:58 hi.txt
```
vs. `lsl`:
```
rclone lsl a                            
        5 2023-11-17 01:33:40.782855000 hi copy.txt
        5 2023-11-17 01:30:58.202211000 hi.txt
```

You are right that it could potentially break things if users are scripting with the output and expecting a certain precision (although I'd have guessed they would use `lsjson` instead.)

nielash replied:
I didn't change this, but I can if you think it's best.

Alternately I'm happy to implement what you described (variable precision per Fs), although that would still be a potentially backwards-incompatible change.

ncw replied:
This is the code that I was thinking of

https://github.com/rclone/rclone/blob/2f5685b40562cf764da673543245ddb13c049a4a/fs/operations/lsjson.go#L50

Maybe this is only used for `lsjson`?

I think we should try not to change the output of `lsf` unless it is guarded with an option.

nielash replied:
Got it. Yeah, looks like this only used by `lsjson`, as `lsf` does not use RFC3339. I have guarded it with a flag in the new version, so that the default is unchanged and nothing should break.

ncw commented on cmd/sync/sync.go:
```diff
@@ -3,22 +3,142 @@ package sync
 
 var (
 	createEmptySrcDirs = false
+	combined           = ""
```
I would just make a LoggerOpt here pass that directly to `AddLoggerFlags`

nielash replied:
Made this change.

ncw commented on backend/drive/drive.go:
```diff
@@ -870,6 +870,11 @@ func (f *Fs) getFile(ctx context.Context, ID string, fields googleapi.Field) (in
 	return info, err
 }
 
+// Options returns remote-specific options for reference by commands
```
I don't like this - this breaks the backend encapsulation and will mean you need to `import "github.com/rclone/rclone/backend/drive` to use it which means that you ignore the users backend choices if they are embedding rclone.

The proper way to do this is to add a new method or a new Feature flag.

However I think you can do something different.

Detect whether you ever receive a file with size < 0 before giving your warning with a `sync.Once`.

This will mean that you catch the other cases because files of unknown length can happen in other backends not just Google Drive

- google drive
- google photos
- s3/gcs with Content-encoding Gzip files

I think this PR should be called `bisync: add support files with unknown length, including Google Docs - fixes #5696`

Generally in this commit `s/Google doc/files of unknown size/g;` :-)

nielash replied:
The main reason for not using size is that Directories also have a size of -1 (if using `--create-empty-src-dirs`). But we do have another way of detecting Dirs, so I think actually I can make this work. 

I had also wanted to not bother users with this warning if they're already using `--drive-skip-gdocs` (which we can't otherwise check). But I think you are right that we can assume they are not, if we see non-dir listings with size of -1.

nielash replied:
Made this change.

ncw commented on cmd/bisync/deltas.go:
```diff
@@ -264,6 +264,7 @@ func (b *bisyncRun) applyDeltas(ctx context.Context, ds1, ds2 *deltaSet) (change
 	// check is a read-only operation by its nature, so it's already "dry" in that sense.
 	ctxNew, ciCheck := fs.AddConfig(ctx)
 	ciCheck.DryRun = false
+	ciCheck.NoUpdateModTime = true
```
Please write why in the commit message!  (I can't comment on the commit messages)

nielash replied:
This is made moot by a different change, so I removed it.

ncw commented on backend/local/local.go:
```diff
@@ -1447,6 +1447,10 @@ func cleanRootPath(s string, noUNC bool, enc encoder.MultiEncoder) string {
 		vol := filepath.VolumeName(s)
+		if vol == `\\?` {
+			// `\\?\C:`
+			vol = s[:6]
```
That fix looks like it is an index error waiting to happen! 

Maybe wrap it in `if len(s) >= 6`?

Its a nasty fix - hopefully we can get rid of it soon!

nielash replied:
Will do! (And agreed.)

nielash replied:
Made this change.

ncw commented on cmd/bisync/cmd.go:
```diff
@@ -50,6 +50,12 @@ type Options struct {
 	Resilient             bool
 	TestFn                TestFunc // test-only option, for mocking errors
 	Retries               int
+	GUI                   bool
```
Being able to show status to the user is a great idea.

However I'd like to consider the other operations too like sync etc.

At the moment you can build a pretty good gui with javascript using the rc.

So I think what you should do is take this commit off the series and stash it on a branch, then make an issue for discussion.

nielash replied:
Ok, I'll remove it for now and make an issue for discussion. I do think there are several bisync-specific things that wouldn't make sense sharing a template with `sync` (for instance, the Path1 / Path2 columns with links to their respective listing files.) ([Demo](https://github.com/nielash/rclone/blob/bisync-logger/docs/static/img/bisync_gui_example.png)) It also displays a warning if a bisync was locked out by a prior run -- which is not a thing in sync.

The rc is very nice but I think it's a drawback that the server needs to be running in order to use it, vs. a static HTML file which can live anywhere (even another rclone remote). I also think the rc adds complexity for new users who want to get up and running quickly. But I'll move it to a branch for now for further discussion!

ncw reviewed (commented):
Wow that was a lot of commits! It contains some excellent work - lovely logging features, loads of bisync fixes and features - well done :-)

I'd like to release v1.65 within the next week or so, so how much of this do you want in that?

I have put some comments inline but there isn't anything serious in there. I'm happy to work on re-arranging the flags later. I don't want to merge the GUI now, but I will consider it for later.

So have a think about which commits you want in v1.65

From my point of view I think we should try to get the logging stuff in, and as many bisync fixes as we can.

PS smaller PRs in future please :-)

nielash commented:
Thanks! And sorry about the massive number of commits! 😅  Nearly all of them depend on earlier ones, so I wasn't sure how I could separate them into different PRs. 

I can take out the GUI for now, that's fine. As for the others, I'd like to get in as many as you'll allow! I can make some quick revisions here based on your comments. (I'll add some responses inline.)

nielash commented on fs/operations/operations.go:
```diff
@@ -2231,3 +2269,326 @@ func SkipDestructive(ctx context.Context, subject interface{}, action string) (s
+	return winner
+}
+
+// DirObjInfo implements fs.ObjectInfo so that logger can log dirs
```
There is probably a better way -- but the problem I was trying to solve is that `LoggerFn` takes:
```
type LoggerFn func(ctx context.Context, sigil Sigil, src fs.ObjectInfo, dst fs.ObjectInfo, err error)
```

and `fs.Directory` does not satisfy `fs.ObjectInfo`. `fs.Dir` also does not work... I get:

```
cannot use dir (variable of type "github.com/rclone/rclone/fs".Dir) as "github.com/rclone/rclone/fs".ObjectInfo value in argument to s.logger: 
"github.com/rclone/rclone/fs".Dir does not implement "github.com/rclone/rclone/fs".ObjectInfo (missing method Fs)compiler[InvalidIfaceAssign]
```

nielash replied:
I didn't change this, but I'm happy to if you can point me towards a better solution.

ncw replied:
LoggerFn should probably be taking an `fs.DirEntry` which is a superset of fs.Object and fs.Directory. If you want more information that is object only then you'd do a type assertion to `fs.Object`

nielash replied:
That worked perfectly -- thank you 🙂 Will send a revised version shortly.

nielash replied:
Fixed in revised version.

nielash commented:
@ncw Here's a revision. I think I followed all of your instructions, but let me know if I misunderstood anything! Also, don't hate me but I added a commit to fix integration tests, and two others that were pre-requisites for that. 

I also rebased and resolved conflicts.

nielash commented on fs/operations/operations.go:
```diff
@@ -2231,3 +2269,326 @@ func SkipDestructive(ctx context.Context, subject interface{}, action string) (s
+	Absolute  bool
+}
+
+// AddLoggerFlags adds the logger flags to the cmdFlags command
```
Made this change (in https://github.com/rclone/rclone/pull/7410/commits/354d81aa95cec98cd3d26dd893b4eebc34a7181d)

nielash commented on fs/operations/operations.go:
```diff
@@ -2231,3 +2269,326 @@ func SkipDestructive(ctx context.Context, subject interface{}, action string) (s
 	}
 	return skip
 }
+
```
Agreed! Made this change.

acegene commented:
> bisync: merge copies and deletes, support --track-renames and --backup-dir -- fixes https://github.com/rclone/rclone/issues/5690 and https://github.com/rclone/rclone/issues/5685

Regarding mostly the above (and to a lesser extent any bisync changes), I am trying to beta test. So far im using the version specified in [tag/v1.65.0-beta.7496.004c12643.bisync-logger-alpha](https://github.com/nielash/rclone/releases/tag/v1.65.0-beta.7496.004c12643.bisync-logger-alpha), any chance you could generate a more up-to-date delivery? Regardless, thanks for your work!

nielash commented:
@acegene thanks! If you're comfortable [installing from source](https://rclone.org/install/#source), you can [download](https://github.com/nielash/rclone/archive/refs/heads/bisync-logger.zip) the latest version from the [nielash:bisync-logger](https://github.com/nielash/rclone/tree/bisync-logger) branch and compile it. Otherwise I'm happy to upload some updated binaries later today or tomorrow!

nielash commented:
@acegene Here are some updated binaries if you'd like to test it out. Feedback is welcome :) 
[Bisync Feature Preview 2023-11-23](https://github.com/nielash/rclone/releases/tag/v1.65.0-beta.7552.7faf0e787.bisync-logger-alpha-2)

crocinsocks commented:
@nielash you are a legend, thank you!! Normal sync without changes takes around 30s whereas before it could take anywhere between 10-20 minutes.. Been using for a few days now without issue.

Reading the comments of --resilient changes I wasn't sure whether this would mean that an exited sync would be able to recover? From my tests it does not however I not using the changes from 3 days ago rather [bisync-logger-alpha-2](https://github.com/nielash/rclone/tree/v1.65.0-beta.7552.7faf0e787.bisync-logger-alpha-2 ). Am I wrong in thinking this and therefore should I open an issue specifically for this case? n.b. this is so a computer can be shutdown / rclone exited without breaking and requiring a --resync..

Also, just off our chats in the past should I open and start an issue for conflict renaming so it is not so ambiguous?

Thank you  🙏

nielash commented:
> @nielash you are a legend, thank you!! Normal sync without changes takes around 30s whereas before it could take anywhere between 10-20 minutes.. Been using for a few days now without issue.

So glad to hear it is working well for you! 😃 

> Reading the comments of --resilient changes I wasn't sure whether this would mean that an exited sync would be able to recover?

It depends on the reason for the exit. `--resilient` allows retrying when bisync has chosen to abort itself due to safety features such as failing `--check-access` or detecting a filter change. It does not currently cover external interruptions such as a user shutting down their computer in the middle of a sync. That said, it's something I would like to add support for, so yes, please open an issue!

> Also, just off our chats in the past should I open and start an issue for conflict renaming so it is not so ambiguous?

I was actually just sketching out a plan for this the other day! I'll open an issue today and tag you so we can discuss it.

nielash commented:
Hi @ncw, question for you -- I am working on some other bisync issues now and would like to submit a new PR for them since this one is too big already 😅  However, they build on the work I did here, so I'm not sure if there's a way to avoid duplicating all of these commits, as I don't think GH supports "dependent" PRs yet. Is there a better way to handle this?

acegene commented:
> Hi @ncw, question for you -- I am working on some other bisync issues now and would like to submit a new PR for them since this one is too big already 😅 However, they build on the work I did here, so I'm not sure if there's a way to avoid duplicating all of these commits, as I don't think GH supports "dependent" PRs yet. Is there a better way to handle this?

I could help with this point. Take a look at [this SO answer](https://stackoverflow.com/a/70178279). Essentially, if you start from this PR/branch (lets call it B1), you could create B2 that depends on it. Locally you could create the branch with:
```
git fetch
git checkout origin/B1
git checkout -b B2 # creates B2 with all the changes that have been pushed to B1
## Add any work you want to B2, then push upstream. then select B1 to merge into so it says "<AUTHOR> wants to merge X commits into B1 from B2".
```

Notes:
- If you have already started the work of B2 on some other branch, lets call it B2_beta, after creating B2 using the above, you could try to cherry-pick the new commits from B2_beta (those not present on B1) onto B2
- You know the process has been successful if the "duplicated" commits have matching commit hashes on each branch
- If new commits are pushed to B1, that you want in B2, you will need to do a rebase of B2 onto origin/B1

I am quite familiar with this type of process, let me know if you have questions or would like to hop in a discord call or something similar.

ncw reviewed (approved):
Wow this is a monster :-)

I think we should merge it now so we can unblock the dev process. We've got time to fix up breakage before the next release.

I'll approve it and you can merge it!

Thank you and many apologies for the backlog.

nielash commented:
Awesome, thank you!! I will do the honors shortly and slay this "monster" 😆
